app/models/order.rb

```
@@ -1,33 +1,29 @@
 class Order < ApplicationRecord
-  STATES = [
-    PENDING = 'pending',
-    ABANDONED = 'abandoned',
-    # Check-out complete; payment authorized.
-    # Buyer credit card has been authorized and hold has been placed.
-    # At this point, availability must be confirmed manually.
-    # Holds expire 7 days after being placed.
-    SUBMITTED = 'submitted',
-    # Availability has been manually confirmed and hold has been "captured" (debited).
-    APPROVED = 'approved',
-    # Items have been deemed unavailable and hold is voided.
-    REJECTED = 'rejected'
-  ]
-
   has_many :line_items, class_name: 'LineItem'
 
-  validates :state, inclusion: { in: STATES }
+  validates :state, presence: true, inclusion: Util::OrderStates.valid_states
 
   before_create :set_code
   before_save :set_currency_code
 
-  scope :pending, -> { where(state: PENDING) }
+  scope :pending, -> { where(state: Util::OrderStates::PENDING) }
 
-  STATES.each do |state|
+  Util::OrderStates.valid_states.each do |state|
     define_method "#{state}?" do
       self.state == state
     end
   end
 
+  Util::OrderStates.events.each do |event|
+    define_method "may_#{event}?" do
+      state_machine.trigger? event
+    end
+
+    define_method "#{event}" do
+      state_machine.trigger event
```

We may consider `trigger!` and raise exceptions, instead of failing silently, when events are triggered from incompatible states.

---

Yeah this might even be a safer functionality to start with unless we'll be rigorously checking for `false` return values and rendering error responses.